# 05 — Model Training

**Plan (agreed with the user):** build the CV/scoring scaffolding once, then
compare models one at a time through that exact same harness — CatBoost
first, then LightGBM, then XGBoost, then a logistic-regression sanity check —
before deciding on reweighting/refinement for whichever comes out on top.

**Why this order, and why shared scaffolding first:** if every model got its
own hyperparameters, its own fold split, or its own metric computation, a
"win" might just mean "we tuned this one more carefully," not "it
generalizes better." So sections 1-2 below (fold assignment, composite
metric, adversarial-OOF probabilities) are built once and imported by every
model section from here on — each new model is a self-contained addition to
this notebook, not a rebuild.

**Why CatBoost first:** its ordered-boosting scheme specifically counters a
form of overfitting where each tree's gradient is computed using predictions
from a model that already saw that same point — a bias that's most
punishing on the kind of moderate-sized, shift-prone data we have here
(1,821 independent locations, 0.984 adversarial-validation AUC from
notebook 04). It's a genuinely relevant first candidate, not an arbitrary
pick.

In [1]:
import sys
sys.path.append('..')

import numpy as np
import pandas as pd
from sklearn.metrics import f1_score, roc_auc_score
from scipy import stats

from src.config import PROCESSED_DIR, MODELS_DIR, SUBMISSIONS_DIR, RANDOM_SEED
from src.features import curated_feature_columns
from src.cv import assign_origin_folds, composite_score, compute_adversarial_oof

pd.set_option('display.max_columns', 30)

train_features = pd.read_csv(PROCESSED_DIR / 'train_features.csv')
test_features = pd.read_csv(PROCESSED_DIR / 'test_features.csv')
print(train_features.shape, test_features.shape)

(18210, 70) (1030, 68)


## 1. Feature set (curated, per notebook 04's evidence)

In [2]:
feature_cols = curated_feature_columns(train_features)
print(f"{len(feature_cols)} features (dropped raw blue_*/red_* aggregates; indices built from them are kept)")
feature_cols

59 features (dropped raw blue_*/red_* aggregates; indices built from them are kept)


['VH_mean',
 'VH_std',
 'VH_min',
 'VH_max',
 'VV_mean',
 'VV_std',
 'VV_min',
 'VV_max',
 'green_mean',
 'green_std',
 'green_min',
 'green_max',
 're1_mean',
 're1_std',
 're1_min',
 're1_max',
 're2_mean',
 're2_std',
 're2_min',
 're2_max',
 're3_mean',
 're3_std',
 're3_min',
 're3_max',
 'nir_mean',
 'nir_std',
 'nir_min',
 'nir_max',
 'nira_mean',
 'nira_std',
 'nira_min',
 'nira_max',
 'swir1_mean',
 'swir1_std',
 'swir1_min',
 'swir1_max',
 'swir2_mean',
 'swir2_std',
 'swir2_min',
 'swir2_max',
 'ndwi_mean',
 'ndwi_std',
 'ndwi_min',
 'ndwi_max',
 'mndwi_mean',
 'mndwi_std',
 'mndwi_min',
 'mndwi_max',
 'ndvi_mean',
 'ndvi_std',
 'ndvi_min',
 'ndvi_max',
 'vh_minus_vv_mean',
 'vh_minus_vv_std',
 'vh_minus_vv_min',
 'vh_minus_vv_max',
 'n_valid_s1_months',
 'n_valid_s2_months',
 's2_valid_fraction_of_s1']

## 2. Fold assignment (origin-level, stratified, shared by every model below)

In [3]:
N_SPLITS = 5
train_features = assign_origin_folds(train_features, n_splits=N_SPLITS, seed=RANDOM_SEED)

# sanity: every origin's variants should all share one fold, and label
# balance per fold should be close to the overall ~40%
fold_check = train_features.groupby('fold').agg(
    n_rows=('ID', 'size'),
    n_origins=('origin_id', 'nunique'),
    positive_rate=('label', 'mean'),
)
fold_check

,n_rows,n_origins,positive_rate
fold,,,
0,3650,365,0.402740
1,3640,364,0.403846
2,3640,364,0.403846
3,3640,364,0.403846
4,3640,364,0.403846


**Why this matters:** if `n_origins` didn't divide evenly (~364/fold) or
`positive_rate` swung far from ~0.40, the fold split itself would be
introducing noise into every model comparison that follows. This confirms
the split is doing its job before we train anything on top of it.

## 3. Adversarial OOF probabilities (shared diagnostic + future reweighting input)

In [4]:
adversarial_oof = compute_adversarial_oof(train_features, test_features, feature_cols, seed=RANDOM_SEED)
train_features['adversarial_prob'] = adversarial_oof

print("Adversarial OOF P(is_test) summary:")
print(train_features['adversarial_prob'].describe())

Adversarial OOF P(is_test) summary:
count    18210.000000
mean         0.058984
std          0.144537
min          0.000073
25%          0.002149
50%          0.007269
75%          0.034177
max          0.994221
Name: adversarial_prob, dtype: float64


**Why compute this now, before any real model:** we'll use it twice —
first as a diagnostic (does a model's error concentrate where this
probability is high?), and later as the sample weight for the reweighting
experiment. Computing it once here, aligned to the exact same folds as every
model below, means both uses stay consistent.

## 4. CatBoost

Conservative-ish settings to start (moderate depth, mild L2 bump, slow
learning rate with generous early stopping) — favoring stability over
squeezing out CV score, per the plan. Training objective/early-stopping uses
built-in AUC rather than a custom composite metric wired into CatBoost's
training loop (simpler and robust across libraries); the real composite
score is then computed directly from each fold's validation predictions.

In [5]:
from catboost import CatBoostClassifier

catboost_oof = np.zeros(len(train_features))
catboost_test_preds = np.zeros((N_SPLITS, len(test_features)))
catboost_fold_scores = []

for fold_i in range(N_SPLITS):
    tr_mask = (train_features['fold'] != fold_i).values
    val_mask = (train_features['fold'] == fold_i).values

    X_tr, y_tr = train_features.loc[tr_mask, feature_cols], train_features.loc[tr_mask, 'label']
    X_val, y_val = train_features.loc[val_mask, feature_cols], train_features.loc[val_mask, 'label']

    model = CatBoostClassifier(
        iterations=3000, learning_rate=0.03, depth=6, l2_leaf_reg=5,
        eval_metric='AUC', random_seed=RANDOM_SEED,
        early_stopping_rounds=150, verbose=False,
    )
    model.fit(X_tr, y_tr, eval_set=(X_val, y_val))

    val_pred = model.predict_proba(X_val)[:, 1]
    catboost_oof[val_mask] = val_pred
    catboost_test_preds[fold_i] = model.predict_proba(test_features[feature_cols])[:, 1]

    f1, auc, comp = composite_score(y_val, val_pred)
    catboost_fold_scores.append({'fold': fold_i, 'f1': f1, 'auc': auc, 'composite': comp,
                                   'best_iteration': model.get_best_iteration()})
    print(f"fold {fold_i}: F1={f1:.4f}  AUC={auc:.4f}  composite={comp:.4f}  (best_iter={model.get_best_iteration()})")

catboost_scores_df = pd.DataFrame(catboost_fold_scores)
print()
print(catboost_scores_df[['f1', 'auc', 'composite']].mean())

fold 0: F1=0.9558  AUC=0.9931  composite=0.9707  (best_iter=1368)


fold 1: F1=0.9639  AUC=0.9937  composite=0.9758  (best_iter=2185)


fold 2: F1=0.9626  AUC=0.9908  composite=0.9739  (best_iter=1077)


fold 3: F1=0.9583  AUC=0.9922  composite=0.9718  (best_iter=1391)


fold 4: F1=0.9766  AUC=0.9974  composite=0.9849  (best_iter=1086)

f1           0.963436
auc          0.993430
composite    0.975434
dtype: float64


## 5. Overall OOF score (all folds combined, not just fold-averaged)

In [6]:
# Fold-averaged and pooled-OOF scores can differ slightly; report both.
f1_oof, auc_oof, comp_oof = composite_score(train_features['label'], catboost_oof)
print(f"Pooled OOF:      F1={f1_oof:.4f}  AUC={auc_oof:.4f}  composite={comp_oof:.4f}")
print(f"Fold-averaged:   F1={catboost_scores_df.f1.mean():.4f}  AUC={catboost_scores_df.auc.mean():.4f}  composite={catboost_scores_df.composite.mean():.4f}")

Pooled OOF:      F1=0.9634  AUC=0.9932  composite=0.9754
Fold-averaged:   F1=0.9634  AUC=0.9934  composite=0.9754


## 6. Diagnostic: do errors concentrate where adversarial shift is highest?

In [7]:
oof_error = np.abs(train_features['label'] - catboost_oof)

corr, pval = stats.spearmanr(train_features['adversarial_prob'], oof_error)
print(f"Spearman correlation (adversarial P(is_test) vs. OOF error): {corr:.4f} (p={pval:.4g})")

# bucket into quartiles of adversarial probability, look at mean error per bucket
train_features['adv_quartile'] = pd.qcut(train_features['adversarial_prob'], 4, labels=['Q1 (least test-like)', 'Q2', 'Q3', 'Q4 (most test-like)'])
train_features['oof_error'] = oof_error
train_features.groupby('adv_quartile', observed=True)['oof_error'].mean()

Spearman correlation (adversarial P(is_test) vs. OOF error): 0.1954 (p=4.321e-156)


adv_quartile
Q1 (least test-like)    0.024369
Q2                      0.033592
Q3                      0.040633
Q4 (most test-like)     0.062701
Name: oof_error, dtype: float64

**How to read this:** a positive correlation, and rising mean error from
Q1 to Q4, would mean CatBoost genuinely struggles more on training rows that
look most "test-like" — direct evidence that the train/test shift is hurting
real predictive accuracy, not just an abstract distribution-comparison
number. That's exactly the situation reweighting is meant to help with.

## 7. Save OOF/test predictions and update the running model-comparison table

In [8]:
oof_path = PROCESSED_DIR / 'oof_predictions.csv'
test_pred_path = PROCESSED_DIR / 'test_predictions.csv'
results_path = PROCESSED_DIR / 'model_results.csv'

# OOF predictions store: one column per model, keyed by ID, so later
# notebook sections/models just add a column rather than rebuilding this.
if oof_path.exists():
    oof_store = pd.read_csv(oof_path)
else:
    oof_store = train_features[['ID', 'label', 'origin_id', 'fold', 'adversarial_prob']].copy()
oof_store['catboost_oof'] = catboost_oof
oof_store.to_csv(oof_path, index=False)

if test_pred_path.exists():
    test_pred_store = pd.read_csv(test_pred_path)
else:
    test_pred_store = test_features[['ID']].copy()
test_pred_store['catboost_pred'] = catboost_test_preds.mean(axis=0)
test_pred_store.to_csv(test_pred_path, index=False)

# running leaderboard across models, appended to as we add each one
new_result = pd.DataFrame([{
    'model': 'catboost',
    'f1_oof': f1_oof, 'auc_oof': auc_oof, 'composite_oof': comp_oof,
    'composite_fold_mean': catboost_scores_df.composite.mean(),
    'composite_fold_std': catboost_scores_df.composite.std(),
}])
if results_path.exists():
    results_store = pd.read_csv(results_path)
    results_store = results_store[results_store['model'] != 'catboost']  # allow re-running this cell
    results_store = pd.concat([results_store, new_result], ignore_index=True)
else:
    results_store = new_result
results_store.to_csv(results_path, index=False)

print("saved:", oof_path, test_pred_path, results_path)
results_store

saved: /home/claude/project/data/processed/oof_predictions.csv /home/claude/project/data/processed/test_predictions.csv /home/claude/project/data/processed/model_results.csv


,model,f1_oof,auc_oof,composite_oof,composite_fold_mean,composite_fold_std
0,catboost,0.963432,0.993242,0.975356,0.975434,0.005642


## Summary of findings (actual results)

- Feature set: **59 curated features** (16 band/index groups x mean/std/min/max,
  minus the 2 dropped groups (blue, red) x 4 stats = 8 fewer than the full 67,
  plus 3 valid-month-count features).
- Fold assignment verified: ~364 origins/fold, positive rate stable at
  40.3-40.4% across all 5 folds — clean split.
- Adversarial OOF probabilities: median P(is_test) is low (0.007) but the
  distribution has real spread (max 0.994) — most training rows look
  train-like, but a meaningful minority look genuinely test-like.
- **CatBoost pooled-OOF: F1 = 0.9634, AUC = 0.9932, composite = 0.9754.**
  Fold-to-fold std of composite is only 0.0056 — consistent across folds.
- **Shift diagnostic confirms the concern is real, not just theoretical:**
  Spearman correlation between adversarial P(is_test) and OOF error is 0.195
  (p < 1e-150) — small in magnitude but highly significant given the sample
  size, and the quartile breakdown is the more telling number: mean error
  rises from 0.024 (Q1, least test-like) to 0.063 (Q4, most test-like) — a
  **~2.6x jump**. CatBoost genuinely struggles more on training rows that
  resemble test, which is direct evidence (not just an abstract distribution
  comparison) that reweighting toward test-like rows is worth trying.
- **Important expectation-setting:** this 0.975 composite is a within-train-
  distribution CV score. Given the 0.98+ adversarial-validation AUC from
  notebook 04, the real leaderboard score is very likely to be meaningfully
  lower than this — CV here measures "how well does the model fit locations
  drawn from train's period," not "how well will it do on test's period."
  Treat 0.975 as an upper bound / model-comparison tool, not a leaderboard
  prediction.
- Saved: `oof_predictions.csv`, `test_predictions.csv` (each with a
  `catboost_*` column, ready for more models to add their own columns),
  `model_results.csv` (running comparison table, currently 1 row).

**Next:** add a LightGBM section using the identical `feature_cols`, `fold`
column, and `composite_score` function, append its row to the same three
files, then compare CatBoost vs. LightGBM head-to-head.